Fix #4119, #4120, and part of #59: Upgrade to Kotlin 1.6.10

BenHenning opened the pull request:
## Explanation
Fixes #4119
Fixes #4120
Fixes part of #59

This PR finishes the migration of the codebase to Kotlin 1.6 (addressing both #4119 and #4120).

Kotlin 1.6 is needed as part of moving rules_kotlin to 1.7.x (which is, in turn, needed in conjunction with Bazel 6.x to enable strict dependency checking which significantly simplifies modularization which is planned for downstream PRs). This PR doesn't actually finish the movement to that version of rules_kotlin, but it does finish moving the codebase to a new enough (and no longer pre-release) version of rules_kotlin to allow using Kotlin 1.6 (over Kotlin 1.4 that the codebase currently uses): version 1.5.0.

Previous PRs (#5400 and #5402) prepared for the changes here by addressing large categories of build warnings that have either arisen from this migration, or from past work. Note that another large category of warnings have also been addressed in this PR: by moving to Kotlin 1.6, there's no longer a runtime incompatibility between the Kotlin SDK and the reflection APIs (which was causing a _lot_ of warning output previously). Between all three PRs, the output is now very clean and free of nearly all build warnings.

To try and keep the warnings clean long-term, this PR introduces warnings-as-errors for both Java and Kotlin code. However, please note some caveats:
- Dagger generated code doesn't follow the Java warnings-as-error flag, so those warnings were cleaned up manually (and will need to be generally watched for, unfortunately).
- The version of rules_kotlin used in this PR doesn't directly support turning on the functionality, but does internally (so a small patch file has been added to augment rules_kotlin). When the codebase is updated to rules_kotlin 1.7.x this patch will no longer be needed.
- To ease development, a build configuration flag was added to disable failure upon encountering build warnings (per https://bazel.build/run/bazelrc and https://bazel.build/docs/configurable-attributes#query-and-cquery as an example), though this needs to be opted into:
  ```sh
  bazel build --config=ignore_build_warnings <target>
  ```

Some other details to note:
- Version 1.6.10 is specifically picked in order to ensure Jetpack Compose compatibility (for preparation of the work being prototyped in #5401 to be compatible with the Oppia Android build environment).
- The vast majority of code in this PR is updating parameterized tests to use a cleaner repeatable annotation pattern that wasn't available in Kotlin 1.4.
- This upgrade absolutely does have runtime implications, but we're relying very heavily on existing automated tests to ensure correctness and no regressions.
- This PR doesn't make an effort to move toward newer Kotlin language features except where forced (API deprecations) or largely wanted (the repeatable annotation change).
- android-spotlight and kotlitex have been updated to support newer versions of Kotlin (as both are custom forks managed in the broader Oppia GitHub organization).
- Gradle files have been updated to match the same dependency versions as Bazel (where it was obvious to make changes; some might still be a bit off).
- The Gradle build configuration was also updated to use Kotlin 1.6.x (otherwise there would be build incompatibilities with Bazel). I think this is the last upgrade we can do for Gradle without upgrading AGP (which will cause us significant issues with the model module, so we're planning on instead dropping Gradle support).
- API changes that needed to be addressed in this PR due to deprecations include: ``String.captialize``, ``String.toLowerCase``, ``String.toUpperCase``, ``SendChannel.offer``, and ``Char.toInt``.
- New API changes that have been leveraged in this PR: ``Flow.lastOrNull`` and ``Deferred.asListenableFuture`` (to replace ``SettableFuture`` for safety; this also resulted in nice simplifications in ``CoroutineExecutorService``).
- The JVM coroutines dependency needed to be split out from Maven and manually imported with some empty internal Java class files since it otherwise has some issues being desugared: https://github.com/bazelbuild/bazel/issues/13553. This is a problem with the Desugarer used in Bazel 4.x (and maybe later versions, so this solution will probably need to kept for a while).
- Some Proguard rule updates were needed due to Kotlin SDK changes--see the Proguard file & comments for specifics.
- Due to dependency changes, the KitKat main dex file was also trimmed down. I'm fairly certain that it's already crashing on startup, so I don't care much about this change--it just needs to build. We plan to remove KitKat entirely eventually, anyway: #5012.
- Jetifier (that is, automatic conversion from support libraries to Jetpack/AndroidX) support was disabled in Gradle. We don't have it enabled in Bazel, and it could potentially encourage strange one-version violations if it was ever actually needed. This is a safer (and likely more performant) change to make.
- Moshi was updated to 1.13 to support the upgrade in Kotlin. This did result in a small configuration change due to its annotation processor being moved. Note that Moshi 1.14 couldn't be supported since it requires Kotlin 1.7+ which requires rules_kotlin 1.7+. This will be an option to upgrade in the future.
- Some improvements and fixes were made in ``FilterPerLanguageResources`` (I think it was outputting something incorrectly before and that's now been fixed as part of a broader logical reworking of the filtering logic).
- ``com.android.support:support-annotation`` was removed as a dependency since it was never used in Bazel, and shouldn't be used (since it's support library and not AndroidX).
- The updates to Moshi and Kotlin dependencies resulted in a bunch of other transitive dependency updates.
- Note that Gradle doesn't have ``allWarningsAsErrors`` enabled since it would require fixing more warnings than is exposed in Bazel, and we're using Bazel builds as the general source of truth for code quality.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This is an infrastructural change. While it could inadvertently affect user-facing code, it shouldn't based on the current passing state of automated tests.

BenHenning committed:
Migrate to rules_jvm_external 5.1.

BenHenning committed:
Fix JVm dep build warnings.

BenHenning committed:
Fix Maven license check & update the list.

The check & related scripts required fairly involved reworking since the
Maven install file (from which it sources Maven URL context) changed in
format as part of the upgrade for rules_jvm_external. This has actually
led to what seems to be more correct analysis of libraries that the
build depends on, so more licenses have been added to the
maven_dependencies.textproto tracking file.

One unused Crashlytics dependency was removed since it was referencing
an old license that doesn't exist anymore (and the artifact should be
replaced in full by more recent Firebase Crashlytics dependencies that
we are already using).

BenHenning committed:
Fix issues with compute_affected_tests.

This addresses an underlying bug with the command executor that can, in
some cases, break compute_affected_tests. It also refines some of its
internal mechanisms for much better performance on expensive PRs.

It also prepares the base support needed for merge queues, but the
CI workflows aren't being updated in this change.

BenHenning committed:
Use base SHA for computing affected tests.

This prepares for merge queues (but doesn't quite configure the workflow
for them--that will happen in a different PR), and improves how tests
are computed for stacked PRs.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/src/java/org/oppia/android/scripts/common/BUILD.bazel
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesRetriever.kt
	scripts/src/java/org/oppia/android/scripts/maven/GenerateMavenDependenciesList.kt
	scripts/src/javatests/org/oppia/android/scripts/license/MavenDependenciesListCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/license/MavenDependenciesRetrieverTest.kt
	scripts/src/javatests/org/oppia/android/scripts/maven/GenerateMavenDependenciesListTest.kt

BenHenning committed:
Post-merge lint fix.

BenHenning committed:
Fix updated TODOs.

Also, update TODO check script to have nicer output, and support
generating the exemption textproto file for easier updates in the
future.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Fix Dagger structure.

This moves the codebase to using the recommended single top-level Dagger
library rather than replicating it in a bunch of different places.

BenHenning committed:
Upgrade Dagger version.

This is needed for downstream work. It also includes ensuring that Guava
JRE can never be used (since only Android should ever be referenced by
the production app build).

BenHenning committed:
Fix test builds that depend on Dagger generation.

BenHenning committed:
Update Maven dependency licenses.

BenHenning committed:
Fix broken test & binary builds.

BenHenning committed:
Add support for Kotlin 1.6.

There's some cleanup work needed beyond this, but this is the core
change to introduce Kotlin 1.6 support (at least for dev builds).

Moshi needed to be upgraded due to a metadata incompatibility when moving
over to Kotlin 1.6.

BenHenning committed:
Correctly interact with rules_kotlin.

rules_kotlin moved its imports into more structured bzl files to load,
so this updates all references in the codebase to point to the correct
locations (which removes debug warnings that show up on the CLI).

BenHenning committed:
Fix upgrade by downgrading Moshi to 1.13.

Moshi 1.14 pulls in kotlin-reflect 1.7.0 which isn't compatible with the
rules_kotlin version we need for Bazel 4.x. Relatedly, this downgrades
rules_kotlin to 1.5.0, but it fortunately keeps all other changes needed
for 1.7.1 (which will be used in a later PR).

Some code fixes were needed, too, for unknown reasons (since the build
should've been using Kotlin 1.6 before). Either way, these fixes seem
reasonable.

BenHenning committed:
Turn on warnings as errors for Kotlin.

Fixed all warnings that the compiler warned about.

Removed ViewModelProvider & fixed state leaking entirely by moving away
from Jetpack's ViewModel as a base class (since we aren't correctly
using that correctly).

BenHenning committed:
Fix Java build & other warnings.

Enables Java warnings-as-errors, though this doesn't yet apply to
kapt-generated code (such as the code from Dagger), but those warnings
were still manually fixed.

This also fixes a small import warning in a proto file, and warnings
when building oppia_dev_kitkat (by updating the main dex list, but it's
likely the build doesn't work anymore, anyway, and it's hard to test
locally).

BenHenning committed:
Try to fix AndroidX databinding compile race.

There's a race condition when building large numbers of app tests
simultaneously that can lead to build failures. While the CI runs are
resilient now to this failure, it'd be better to try and fix it. This
removes the last non-AndroidX dependency from the codebase with hopes
that it helps reduce the likelihood of the error (though there are no
dependencies on it, so it's unlikely).

BenHenning committed:
Fix static check failures.

BenHenning commented on .bazelrc:
```diff
@@ -6,5 +6,7 @@ build --android_databinding_use_v3_4_args \
     --define=android_standalone_dexing_tool=d8_compat_dx \
     --android_databinding_use_androidx
 
+build --javacopt="-Werror"
```
Add comment explaining this.

BenHenning commented on tools/kotlin/BUILD.bazel:
```diff
@@ -0,0 +1,30 @@
+# This package defines codebase-wide third-party build configurations for Kotlin.
```
Convert to file docstring.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

BenHenning committed:
Fix CI breakages, including Gradle.

BenHenning committed:
Fix broken CI checks.

Specifically:
- The TODO check was fixed by reformatting a TODO comment.
- The Proguard build was fixed by upgrading kotlinx.coroutines to use a
  version compatible with Kotlin 1.6, as well as adding a Proguard
  dontwarn directive for one class that can't execute on Android.

BenHenning committed:
Address self-review comments.

BenHenning committed:
Fix broken tests.

For some reason, the old kotlinx-coroutines-test was causing a
codebase-wide incompatibility with the 1.6.x kotlinx-coroutines-core
library, so this updates that version.

It also updates Gradle dependencies for consistency, and to fix the
Gradle builds.

And, this fixes the IntelliJ Bazel plugin syncing issue when using
kt_jvm_import (for kotlinx-coroutines-core-jvm).

BenHenning committed:
Address TODOs as part of the migration.

This addresses TODOs for #4119 and #4120. It also updates TODOs on #3715
to #4463, instead, since these apparently weren't correctly originally.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Fix breakage from develop.

The new proto target isn't used anywhere so this was missed.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/res/layout-land/profile_chooser_profile_view.xml
	app/src/main/res/layout-sw600dp-port/profile_chooser_profile_view.xml
	app/src/main/res/layout/profile_chooser_profile_view.xml

BenHenning committed:
Fix hanging CoroutineExecutorService test.

I'm not sure of the underlying problem, but we've run into issues with
the custom Deferred-to-Future conversion mechanism. This replaces that
with an existing kotlinx Deferred-to-Guava ListenableFuture mechanism
that is used elsewhere in the codebase and it seems to fix the stability
issue.

BenHenning committed:
Fix broken Gradle build.

BenHenning commented:
Still active.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/test_file_exemptions.textproto
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	app/src/main/java/org/oppia/android/app/translation/BUILD.bazel
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	.github/CODEOWNERS
	app/src/main/java/org/oppia/android/app/options/AppLanguageActivity.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragment.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/OptionsActivity.kt
	app/src/main/java/org/oppia/android/app/options/OptionsActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/options/OptionsFragmentPresenter.kt
	app/src/sharedTest/java/org/oppia/android/app/options/AppLanguageActivityTest.kt
	app/src/sharedTest/java/org/oppia/android/app/options/AppLanguageFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/options/OptionsFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/test/java/org/oppia/android/app/translation/AppLanguageResourceHandlerTest.kt
	app/src/test/java/org/oppia/android/app/utility/math/MathExpressionAccessibilityUtilTest.kt
	scripts/src/java/org/oppia/android/scripts/build/FilterPerLanguageResources.kt
	third_party/maven_install.json
	utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix a variety of dev platform-specific issues.

These issues were found after I started using a new development
environment.

BenHenning committed:
Tidy some things up, and revert fragment test.

ProfileAndDeviceIdFragmentTest had been updated to use a newer fragment
initialization pattern, but that's no longer needed and seems to be
causing what appears to be timing discrepancies between local dev and
CI.

BenHenning committed:
Fix test behavior inconsistency in test.

The issue ultimately arose from test parameters being initialized after
they're needed in the launched UI. This type of change was tried earlier
in the branch, but reverted since it didn't seem necessary. It is,
however, necessary when there are environment differences (e.g. local
vs. CI) or when running certain tests individually.

Due to the difficulty in finding this issue, ActivityScenarioRule has
been added as a prohibited pattern in the static regex checks (along
with ActivityTestRule since that's deprecated and discouraged, anyway).

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Fix broken ProfileAndDeviceIdFragmentTest test.

The test was suffering from some proto encoding inconsistencies that
seem to occur between some development machines vs. on CI. The fix
improves the test's robustness by extracting the raw encoded string,
verifying that the other outputs in the intent message correctly
correspond to that string, and that the string (as a parsed proto)
contains the correct values. As a result, the test no longer depends on
a hardcoded encoding value to be present for verification. This does
result in a bit more logic than is generally good to have in a test (and
it lengthened the test code quite a bit), but it seems necessary in this
particular case.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Post-merge fix.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Some follow-up post-merge fixes.

This doesn't address all timeouts being observed in scripts. More
analysis may be needed.

BenHenning committed:
Make TodoOpenCheck locally runnable.

This does a bunch of other small things, too, but the main difference
here is introducing support for interacting with GitHub via its REST API
rather than requiring the user to use the 'gh' tool locally (though the
gh CLI tool is still needed for maintaining authentication access).

BenHenning committed:
Simplify GitHubIssue & some static fixes/docs.

Tests still need to be added, and some cleanup is needed once the
branch's base is adjusted.

BenHenning committed:
Add ScriptBackgroundCoroutineDispatcher.

This is a script-specific dispatcher which will allow for better
asynchronous support in upcoming PRs (especially for command execution).
This change serves to prepare for those changes.

BenHenning committed:
Some more robustness fixes.

The main change here is ensuring that Bazel 4.0.0 is used & bzlmod
disabled in newer versions of Bazel when running operations in a test
Bazel environment.

This commit also introduces some more timing tweaks on CommandExecutor
for some tests, though these only affect very specific tests (as many
script tests directly call a script's main() function and thus don't
overwrite its executor behavior).

This commit attempted to introduce "--batch" mode to runs, but the
isolation didn't actually seem to improve stability and, instead,
substantially slowed down some of the tests.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

Conflicts:
	scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/BazelClientTest.kt

BenHenning committed:
Post-merge fixes.

Plus, actually makes use of the new script background dispatcher in
CommandExecutorImpl to make the new wiring make more sense (though its
real utility will come in a follow-up PR).

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Post-merge fixes & adjustments.

This makes the TODO open check script runnable using the script
background dispatcher, and fixes some tests so that they pass now.

More documentation and testing work is still needed to finalize this PR.

BenHenning committed:
Fix BUILD file lint issue.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Static fix checks.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt
	scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt
	scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt
	scripts/src/java/org/oppia/android/scripts/common/GitClient.kt
	scripts/src/java/org/oppia/android/scripts/todo/BUILD.bazel
	scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/ScriptBackgroundCoroutineDispatcherTest.kt
	scripts/src/javatests/org/oppia/android/scripts/testing/TestGitRepositoryTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' of github.com:oppia/oppia-android into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/test_file_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Add smoke tests for instr. binaries & tests.

BenHenning committed:
Some minor refactoring for readability.

BenHenning committed:
Fixed broken instrumentation builds.

This was done by removing the //testing dependency and, instead, having
instrumentation targets depend on the direct module within //testing
that they need to build. This module & its corresponding implementation
binding (and tests) needed to be moved out of //testing and into their
own /firebase package.

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Add missing CODEOWNERS line.

BenHenning committed:
Remove old files from CODEOWNERS.

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Add missing tests for TestBlazeWorkspace changes.

Also includes some minor clean-ups within TestBazelWorkspaceTest.

BenHenning committed:
Add missing Firebase auth tests.

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Post-merge fixes for previously missed tests.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Improve new test robustness.

Required upgrade of kotlinx-coroutines-core.

BenHenning committed:
Fix updated Maven license deps.

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Update scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Add tests for GitHubClient & TodoOpenCheck.

Add new FakeCommandExecutor utility + tests.

Add missing GitHub workflow update.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Add missing docs.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/maven_dependencies.textproto
	third_party/maven_install.json

BenHenning committed:
Update Maven dep list checker & updater scripts.

Updates the utilities for the Maven dep list & updater scripts to have
much more robustness when trying to verify the correct repository URLs
for artifact coordinates, and when downloading artifact POM files. The
main robustness changes come in the form of retries and smarter
parallelization and error checking.

These changes were necessary due to one of my workstations frequently
flaking out on large numbers of URL blocks during verification, and this
helps considerably.

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Fix broken TodoOpenCheck tests & check exemptions.

For the former, this mainly involved isolating the gh auth tool
dependency in the same way as GitHubClientTest. This improves
environment independence and test performance.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Fix broken checks & failing static checks.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Includes some manual fixes due to ":dagger" being moved to top-level.

Conflicts:
	app/BUILD.bazel
	app/src/main/java/org/oppia/android/app/notice/testing/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/notice/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/exploration/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/onboarding/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	scripts/assets/maven_dependencies.textproto
	third_party/maven_install.json
	third_party/versions.bzl
	utility/src/main/java/org/oppia/android/util/logging/firebase/BUILD.bazel
	utility/src/test/java/org/oppia/android/util/logging/firebase/BUILD.bazel

BenHenning committed:
Post merge build, test, and static check fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Includes some additional clean-up changes in ClickableAreasImage.

Conflicts:
	app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/testing/InputInteractionViewTestActivity.kt
	app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
	app/src/main/res/layout/activity_input_interaction_view_test.xml
	app/src/main/res/layout/nav_header_navigation_drawer.xml
	app/src/main/res/layout/profile_picture_activity.xml
	domain/build.gradle
	instrumentation/oppia_instrumentation_test.bzl
	scripts/assets/maven_dependencies.textproto
	testing/build.gradle
	third_party/maven_install.json
	third_party/versions.bzl

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	utility/src/main/java/org/oppia/android/util/locale/BUILD.bazel

BenHenning committed:
Some post-merge fixes.

More are needed yet to address all of the Kotlin warnings (which are now
treated as errors).

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

BenHenning committed:
Silence deprecation warnings.

These will need to be fixed properly before the PR is ready for review.

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Address reviewer comment.

seanlip committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge commit 'e2f94e41e4cba3bc5b5f3af65f7c43a6543b4c52' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	WORKSPACE
	third_party/maven_install.json

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	third_party/maven_install.json
	utility/src/test/java/org/oppia/android/util/logging/BUILD.bazel

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/java/org/oppia/android/app/testing/InputInteractionViewTestActivity.kt
	app/src/main/res/layout/activity_input_interaction_view_test.xml
	app/src/sharedTest/java/org/oppia/android/app/customview/interaction/MathExpressionInteractionsViewTest.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/AnalyticsController.kt
	third_party/maven_install.json

BenHenning committed:
Address self-review comments.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Address review comment.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	domain/src/main/java/org/oppia/android/domain/auth/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/oppialogger/survey/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/survey/BUILD.bazel
	testing/src/main/java/org/oppia/android/testing/firebase/BUILD.bazel
	third_party/maven_install.json

BenHenning committed:
Fix NetworkLoggingInterceptorTest.

BenHenning committed:
Fix more tests & upgrade Robolectric.

Fixing one of the test warnings required upgrading to Robolectric 4.9
(to add support for SDK 33 which the app has targeted for the past ~6
months).

BenHenning committed:
Fix more broken tests.

BenHenning committed:
Undo Robolectric upgrade.

Robolectric upgrade is hard: it requires upgrading Espresso in order to
work around https://github.com/robolectric/robolectric/issues/7367 and
https://github.com/robolectric/robolectric/issues/6593. However, newer
Espresso makes use of new manifest features that require upgrading
Bazel: https://github.com/bazelbuild/bazel/issues/14285. Because
upgrading Kotlin is a precursor to the Bazel upgrade, a workaround for
the single test that required the Robolectric upgrade (due to using a
deprecated API from SDK 33 without being able to actually use SDK 33 in
the test) is the best approach here until Robolectric can actually be
upgraded in the future.

BenHenning committed:
Upgrade to rules_kotlin 1.5.0 beta 3.

This is in preparation for the codebase-wide upgrade to Kotlin 1.6.

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into upgrade-to-kotlin1.6

Conflicts:
	WORKSPACE
	scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
	third_party/versions.bzl

BenHenning committed:
Fix build-time Kotlin & Java warnings.

This pulls in non-configuration changes from #4937 that are compatible with the
current Kotlin 1.4.x. Kotlin 1.5 and 1.6 changes will need to come as part of a
downstream PR.

BenHenning committed:
Fix a few follow-up issues.

These are mainly reverting a few more Kotlin 1.5/1.6-specific API
changes.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/customview/interaction/MathExpressionInteractionsViewTest.kt
	app/src/test/java/org/oppia/android/app/translation/AppLanguageResourceHandlerTest.kt
	data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt
	testing/src/main/java/org/oppia/android/testing/junit/ParameterizedMethod.kt
	utility/src/test/java/org/oppia/android/util/logging/ConsoleLoggerTest.kt

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Follow-up variable rename for clarity.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Address static check failures.

This mainly involved updating CODEOWNERS & fixing a bunch of TODO
references for now-marked deprecations.

BenHenning committed:
Move forward TODO fixes from #4937.

These changes should have been in this branch originally.

BenHenning committed:
Remove unused view model bridge classes.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/java/org/oppia/android/app/viewmodel/BUILD.bazel

BenHenning reviewed (commented):
Self-reviewed all changes.

BenHenning commented:
PTAL @seanlip for CODEOWNERS & other OWNERS changes.

PTAL @adhiamboperes. Note that you may want to review the tests in isolation. They make up the very large majority of changes in this PR, and are largely mechanical (due to making some small formatting and syntax changes to move over the ``Iteration`` annotation code), but actually reviewing it may take some time as more than 2,000 iteration lines have been impacted.

seanlip reviewed (approved):
LGTM for 4 codeowner files

BenHenning committed:
Pull in other warning fixes from #4886.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Pulled forward Gradle change from #4886.

BenHenning reviewed (commented):
Self-review latest change (pulled in a Gradle file update from #4886).

BenHenning committed:
Merge branch 'develop' into update-rules-jvm-external

BenHenning committed:
Merge branch 'develop' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'develop' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Down version to 1.6.10.

This is to ensure Jetpack Compose compatibility.

BenHenning reviewed (commented):
Self-reviewed latest changes.

FYI I downgraded Kotlin to 1.6.10 for Jetpack Compose support (per the findings in #5401).

BenHenning committed:
Merge branch 'develop' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning commented on tools/kotlin/BUILD.bazel:
```diff
@@ -0,0 +1,32 @@
+
+kt_kotlinc_options(
+    name = "oppia_kotlinc_options",
+    warn = "error",
```
Reminder to update this to be 'report' for local builds and 'error' for CI builds.

adhiamboperes replied:
Is this to be done as part of this PR?

BenHenning replied:
Yep, will be included in the latest batch of changes.

BenHenning committed:
Address review comment.

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

Conflicts:
	domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsControllerImpl.kt

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	domain/BUILD.bazel
	third_party/versions.bzl

BenHenning reviewed (commented):
Pulled in upstream changes and self-reviewed. No concerns at the moment.

adhiamboperes commented on testing/src/main/java/org/oppia/android/testing/junit/OppiaParameterizedTestRunner.kt:
```diff
@@ -73,10 +70,9 @@ import kotlin.reflect.KClass
- * value via [Iteration]).
+ * annotating with [Iteration]. Such tests should not ever read from the [Parameter]-annotated
+ * fields since there's no way to ensure what values those fields will contain (thus they should be
+ * treated as undefined outside of tests that specific define their value via [Iteration]).
```
Nit:
```suggestion
 * treated as undefined outside of tests that specifically define their value via [Iteration]).
```

BenHenning replied:
Done in latest.

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/customview/interaction/MathExpressionInteractionsViewTest.kt:
```diff
@@ -459,40 +450,38 @@ class MathExpressionInteractionsViewTest {
-      "text=(x^2-1)/(x+1)=y/2",
-      "expLatex=(x ^ {2} - 1) \\div (x + 1) = y \\div 2"
-    )
+  @Iteration("numeric_-1", "type=NUMERIC_EXPRESSION", "text=  -  1           ", "expLatex=-1")
```
For tests with these large spaces in the parameters, are they intentional? Ditto [MathExpressionAccessibilityUtilTest.kt](https://github.com/oppia/oppia-android/pull/4937/files#diff-576b417a50621ba1b2b450deb862de133b43108cf491b587a874eae5df6f57e4) below.

BenHenning replied:
Yep, this is intentional. It's partly verifying that arbitrary spaces in inputs don't cause any issues with the math parsers used to analyze these raw expressions.

adhiamboperes commented on domain/src/test/java/org/oppia/android/domain/classify/rules/numericexpressioninput/NumericExpressionInputMatchesUpToTrivialManipulationsRuleClassifierProviderTest.kt:
```diff
@@ -76,19 +73,17 @@ class NumericExpressionInputMatchesUpToTrivialManipulationsRuleClassifierProvide
-  )
+  @Iteration("-2==-2", "answer=-2", "input=-2")
+  @Iteration("1+3.14==1+3.14", "answer=1+3.14", "input=1+3.14")
+  @Iteration(" 1 +   3.14 ==1+3.14", "answer= 1 +   3.14 ", "input=1+3.14")
```
Same question about spaces.

BenHenning replied:
Yep it's intentional--see comment above.

adhiamboperes reviewed (commented):
Thanks @BenHenning!

Apart from a few clarifications and a nit, everything looks good.

I did also wonder about the comment regarding `FilterPerLanguageResources`. If the changes fixed an error output, is it testable?

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Add config for Kotlin warnings.

Add a config to allow warnings-as-errors to be disabled for local
development to make things easier. The default is still to treat them as
errors (and that will be the build configuration for CI, so PRs
shouldn't be mergeable with Kotlin build warnings).

BenHenning committed:
Address reviewer comment.

BenHenning reviewed (commented):
Self-reviewed latest changes.

BenHenning commented:
> Thanks @BenHenning!
> 
> Apart from a few clarifications and a nit, everything looks good.
> 
> I did also wonder about the comment regarding `FilterPerLanguageResources`. If the changes fixed an error output, is it testable?

Thanks @adhiamboperes! I've addressed my earlier + your comments in the latest changes, and brought the PR up-to-date with develop. As part of this, I also updated the PR description to explain how to use the new opt-in build flag for disabling warnigns-as-errors. PTAL.

Regarding ``FilterPerLanguageResources``, I went back to look at the changes and tests. It seems to me that the existing test coverage seems pretty good for verifying that the changes still work as expected (and particularly ``testUtility_whenResourcesAreRemoved_printsDiagnosticInformation``). The exact failures that were originally fixed I have, unfortunately, forgotten as this change was made over a year ago and has kind of sat while the rest of the Bazel chain was being worked on. I think the important takeaway is that it's functionally still working. If you think additional testing is important here, could you please let me know where you think the functional gap is? I'm always happy to add more tests.

BenHenning committed:
Address lint failures.

BenHenning reviewed (commented):
Self-reviewed lint fix.

adhiamboperes reviewed (approved):
Thanks @BenHenning! Everything looks good. I'm happy to leave the filter language resources as is.

adhiamboperes commented:
Merging since everything is up to date.

adhiamboperes commented:
1 test is flaky:

```
org.oppia.android.data.backends.gae.NetworkLoggingInterceptorTest > testLoggingInterceptor_makeNetworkCall_succeeds FAILED
    java.lang.IllegalStateException: This job has not completed yet
        at kotlinx.coroutines.JobSupport.getCompletedInternal$kotlinx_coroutines_core(JobSupport.kt:1199)
        at kotlinx.coroutines.DeferredCoroutine.getCompleted(Builders.common.kt:100)
        at org.oppia.android.data.backends.gae.NetworkLoggingInterceptorTest.testLoggingInterceptor_makeNetworkCall_succeeds(NetworkLoggingInterceptorTest.kt:99)
```
